File: src/abc/norm/pure/Term.js
// This file is a part of Denshi.
// Copyright (C) 2019 Matthew Blount

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public
// License along with this program.  If not, see
// <https://www.gnu.org/licenses/.

import assert from "../../../assert.js";

export class Term {
  constructor() {

  }

  static get id()      { return kId      }
  static get apply()   { return kApply   }
  static get bind()    { return kBind    }
  static get copy()    { return kCopy    }
  static get drop()    { return kDrop    }
  static get reset()   { return kReset   }
  static get shift()   { return kShift   }
  static get sum()     { return kSum     }
  static get negate()  { return kNegate  }
  static get product() { return kProduct }
  static get invert()  { return kInvert  }
  static get exp()     { return kExp     }
  static get log()     { return kLog     }
  static get cos()     { return kCos     }
  static get sin()     { return kSin     }
  static get floor()   { return kFloor   }
  static get ceil()    { return kCeil    }
  
  static real(value) { return new Literal("real", value) }
  static hint(name)  { return new Literal("hint", name)  }
  static word(name)  { return new Literal("word", name)  }
  static block(body) { return new Unary("block", body)   }

  static sequence(fst, snd) {
    if (Term.isId(snd)) {
      return fst;
    }
    if (Term.isId(fst)) {
      return snd;
    }
    if (Term.isSequence(fst)) {
      let inner = Term.sequence(fst.snd, snd);
      return Term.sequence(fst.fst, inner);
    }
    return new Binary("sequence", fst, snd);
  }

  static isTerm(obj)     { return obj instanceof Term     }
  static isConstant(obj) { return obj instanceof Constant }
  static isLiteral(obj)  { return obj instanceof Literal  }
  static isUnary(obj)    { return obj instanceof Unary    }
  static isBinary(obj)   { return obj instanceof Binary   }

  static isId(obj) {
    return Term.isConstant(obj) && obj.name === "id";
  }

  static isApply(obj) {
    return Term.isConstant(obj) && obj.name === "apply";
  }

  static isBind(obj) {
    return Term.isConstant(obj) && obj.name === "bind";
  }

  static isCopy(obj) {
    return Term.isConstant(obj) && obj.name === "copy";
  }

  static isDrop(obj) {
    return Term.isConstant(obj) && obj.name === "drop";
  }

  static isReset(obj) {
    return Term.isConstant(obj) && obj.name === "reset";
  }

  static isShift(obj) {
    return Term.isConstant(obj) && obj.name === "shift";
  }

  static isSum(obj) {
    return Term.isConstant(obj) && obj.name === "sum";
  }

  static isNegate(obj) {
    return Term.isConstant(obj) && obj.name === "negate";
  }

  static isProduct(obj) {
    return Term.isConstant(obj) && obj.name === "product";
  }

  static isInvert(obj) {
    return Term.isConstant(obj) && obj.name === "invert";
  }

  static isExp(obj) {
    return Term.isConstant(obj) && obj.name === "exp";
  }

  static isLog(obj) {
    return Term.isConstant(obj) && obj.name === "log";
  }

  static isCos(obj) {
    return Term.isConstant(obj) && obj.name === "cos";
  }

  static isSin(obj) {
    return Term.isConstant(obj) && obj.name === "sin";
  }

  static isFloor(obj) {
    return Term.isConstant(obj) && obj.name === "floor";
  }

  static isCeil(obj) {
    return Term.isConstant(obj) && obj.name === "ceil";
  }

  static isReal(obj) {
    return Term.isLiteral(obj) && obj.type === "real";
  }

  static isHint(obj) {
    return Term.isLiteral(obj) && obj.type === "hint";
  }

  static isWord(obj) {
    return Term.isLiteral(obj) && obj.type === "word";
  }

  static isBlock(obj) {
    return Term.isUnary(obj) && obj.type === "block";
  }

  static isSequence(obj) {
    return Term.isBinary(obj) && obj.type === "sequence";
  }
}

export class Constant extends Term {
  constructor(name) {
    super();
    this.name = name;
  }
}

export class Literal extends Term {
  constructor(type, value) {
    super();
    this.type = type;
    this.value = value;
  }
}

export class Unary extends Term {
  constructor(type, body) {
    super();
    this.type = type;
    this.body = body;
  }
}

export class Binary extends Term {
  constructor(type, fst, snd) {
    super();
    this.type = type;
    this.fst = fst;
    this.snd = snd;
  }
}

export class Real {
  static sum(fst, snd) {
    assert(Term.isReal(fst));
    assert(Term.isReal(snd));
    return Term.real(fst.value + snd.value);
  }

  static negate(obj) {
    assert(Term.isReal(obj));
    return Term.real(0 - obj.value);
  }

  static product(fst, snd) {
    assert(Term.isReal(fst));
    assert(Term.isReal(snd));
    return Term.real(fst.value * snd.value);
  }

  static invert(obj) {
    assert(Term.isReal(obj));
    if (obj.value === 0) {
      return obj;
    } else {
      return Term.real(1 / obj.value);
    }
  }

  static exp(obj) {
    assert(Term.isReal(obj));
    return Term.real(Math.exp(obj.value));
  }

  static log(obj) {
    assert(Term.isReal(obj));
    return Term.real(Math.log(obj.value));
  }

  static cos(obj) {
    assert(Term.isReal(obj));
    return Term.real(Math.cos(obj.value));
  }

  static sin(obj) {
    assert(Term.isReal(obj));
    return Term.real(Math.sin(obj.value));
  }

  static floor(obj) {
    assert(Term.isReal(obj));
    return Term.real(Math.floor(obj.value));
  }

  static ceil(obj) {
    assert(Term.isReal(obj));
    return Term.real(Math.ceil(obj.value));
  }
}

const kId      = new Constant("id");
const kApply   = new Constant("apply");
const kBind    = new Constant("bind");
const kCopy    = new Constant("copy");
const kDrop    = new Constant("drop");
const kReset   = new Constant("reset");
const kShift   = new Constant("shift");
const kSum     = new Constant("sum");
const kNegate  = new Constant("negate");
const kProduct = new Constant("product");
const kInvert  = new Constant("invert");
const kExp     = new Constant("exp");
const kLog     = new Constant("log");
const kCos     = new Constant("cos");
const kSin     = new Constant("sin");
const kFloor   = new Constant("floor");
const kCeil    = new Constant("ceil");

File: src/MagicWord.Lang/Term.fs
// This file is a part of Magic Word.
// Copyright (C) 2019 Matthew Blount

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public
// License along with this program.  If not, see
// <https://www.gnu.org/licenses/.

namespace MagicWord.Lang

// Terms are a more committal level of parsing than words: code is now
// tree shaped and will evaluate by rewriting. But e.g. we could have
// gone to graphs instead, with something like an interaction net.
module Term =
  type Term =
    | Id
    | Apply
    | Bind
    | Copy
    | Drop
    | Inequal
    | Reset
    | Shift
    | Bang of string
    | Symbol of string
    | Variable of string
    | Quote of Term
    | Sequence of (Term * Term)

  let rec sequence fst snd =
    match (fst, snd) with
      | (_, Id)                    -> fst
      | (Id, _)                    -> snd
      | (Sequence (head, tail), _) -> sequence head <| sequence tail snd
      | _                          -> Sequence (fst, snd)

  type private ReaderState = {
    stack : Term list list
    build : Term list
  }

  // Save the current term on the stack and start a new one.  Called
  // when the reader reaches a '['.
  let private save ctx =
    { build = []; stack = ctx.build :: ctx.stack }

  // Try to pop the stack and make a quote from the saved term. If the
  // stack is empty, then the brackets are unbalanced, and we return
  // in the error branch.
  let private tryRestore ctx =
    match ctx.stack with
      | prev :: rest ->
        let build = List.rev ctx.build
        let block = Quote <| List.fold sequence Id build
        Some <| { build = block :: prev; stack = rest }
      | [] ->
        None

  // Put a term on the building stack.
  let private push ctx term =
    { ctx with build = term :: ctx.build }

  // Perform one step of the reader's computation.
  let private stepReader ctx word =
    match ctx with
      | None     -> None
      | Some ctx ->
        match word with
          | Word.Begin         -> Some <| save ctx
          | Word.End           -> tryRestore ctx
          | Word.Id            -> Some ctx
          | Word.Apply         -> Some <| push ctx Apply
          | Word.Bind          -> Some <| push ctx Bind
          | Word.Copy          -> Some <| push ctx Copy
          | Word.Drop          -> Some <| push ctx Drop
          | Word.Inequal       -> Some <| push ctx Inequal
          | Word.Reset         -> Some <| push ctx Reset
          | Word.Shift         -> Some <| push ctx Shift
          | Word.Bang name     -> Some <| push ctx (Bang name)
          | Word.Symbol name   -> Some <| push ctx (Symbol name)
          | Word.Variable name -> Some <| push ctx (Variable name)

  let parse (words: Word list) : Term option =
    let init = Some <| { build = []; stack = []; }
    match List.fold stepReader init words with
      | None -> None
      | Some ctx ->
        let build = List.rev ctx.build
        Some <| List.fold sequence Id build

  let rec quote (term: Term): Word list =
    match term with
      | Id            -> [Word.Id]
      | Apply         -> [Word.Apply]
      | Bind          -> [Word.Bind]
      | Copy          -> [Word.Copy]
      | Drop          -> [Word.Drop]
      | Inequal       -> [Word.Inequal]
      | Reset         -> [Word.Reset]
      | Shift         -> [Word.Shift]
      | Bang name     -> [Word.Bang name]
      | Symbol name   -> [Word.Symbol name]
      | Variable name -> [Word.Variable name]
      | Quote body    ->
        let body = quote body
        List.concat [[Word.Begin]; body; [Word.End]]
      | Sequence (fst, snd) ->
        let fst = quote fst
        let snd = quote snd
        List.concat [fst; snd]

  type private RewriteState = {
    sink : Term list
    data : Term list
    code : Term list
  }

  let private thunk st term =
    let sink' = List.append st.sink st.data
    let sink' = term :: sink'
    { st with sink = sink'; data = [] }

  let rec private fetch st =
    match st.code with
      | Sequence (fst, snd) :: tail ->
        fetch <| { st with code = fst :: snd :: tail }
      | head :: tail ->
        Some <| (head, { st with code = tail })
      | [] -> None

  let rewrite (term: Term): Term =
    term
